server/auth/auth.controller.js

```diff
@@ -0,0 +1,247 @@
+    email,
+    phoneNumber,
+    role,
+    password: hashedPassword
+  });
+
+  const token = crypto.randomBytes(20).toString('hex');
+  const verifyToken = crypto
+    .createHash('sha256')
+    .update(token)
+    .digest('hex');
+  user.verifyAccountToken = verifyToken;
+  await sendMailVerifyAccount(user.username, user.email, verifyToken);
+
+  await user.save();
+  return res.status(201).json({ success: true });
+});
+
+exports.login = asyncHandler(async (req, res, next) => {
+  const { error } = loginVal(req.body);
+  if (error) return next(new ErrorResponse(error.details[0].message, 400));
+
+  const { email, password } = req.body;
+  let user = await AuthModel.findOne({ email }).select('+password');
+  if (!user) return next(new ErrorResponse('User not found', 404));
+
+  const isMatched = await bcrypt.compare(password, user.password);
+  if (isMatched) {
+    if (!user.isAccountActive) {
+      user.isAccountActive = true;
+      await user.save();
+    }
+
+    const accessToken = await jwt.sign(
+      { id: user._id, username: user.username },
+      config.jwtAccessKey,
+      { expiresIn: config.jwtAccessKeyExpireTime }
+    );
+    const refreshToken = await jwt.sign(
+      { id: user._id, username: user.username },
+      config.jwtRefreshKey,
+      { expiresIn: config.jwtRefreshKeyExpireTime }
+    );
+
+    return res.status(200).json({
+      success: true,
+      accessToken,
+      refreshToken
+    });
+  }
+
+  return next(new ErrorResponse('Password not matching', 401));
+});
+
+exports.currentUser = asyncHandler(async (req, res, next) => {
+  return res.status(200).json({ success: true, user: req.user });
+});
+
+// exports.verifyAccountSendToken = asyncHandler(async (req, res, next) => {
+//   const user = await AuthModel.findById(req.body.userID);
+
+//   if (user.isAccountVerified)
+//     return next(new ErrorResponse('Account is already verified', 400));
+
+//   const token = crypto.randomBytes(20).toString('hex');
+//   const verifyToken = crypto
+//     .createHash('sha256')
+//     .update(token)
+//     .digest('hex');
+//   user.verifyAccountToken = verifyToken;
+//   await user.save();
+//   await sendMailVerifyAccount(user.username, user.email, verifyToken);
+//   return res.status(200).json({ success: true, verifyToken });
+// });
+
+exports.verifyAccountReceiveToken = asyncHandler(async (req, res, next) => {
+  const { token } = req.params;
+  const user = await AuthModel.findOne({
+    verifyAccountToken: token
+  });
+  if (!user) return next(new ErrorResponse('User not found', 404));
+  user.verifyAccountToken = undefined;
+  user.isAccountVerified = true;
+  await user.save();
+  return res.status(200).json({ success: true });
+});
+
+exports.forgetPasswordSendToken = asyncHandler(async (req, res, next) => {
+  const { error } = forgetPasswordVal(req.body);
+  if (error) return next(new ErrorResponse(error.details[0].message, 400));
+  const resetToken = crypto.randomBytes(20).toString('hex');
```

Comment:
can you explain this part?

Reply:
Random crypto token !!  https://nodejs.org/api/crypto.html#crypto_crypto_randombytes_size_callback